File: xaiop-sdk/nodejs/test/array.select.history.test.js
/**
 * Protocol 0.7 Draft — `?` select + bare `&` across history / jumpTo / cover /
 * intercept / window / stream. Lockstep with Python, Java, Go.
 */

import assert from "node:assert/strict";
import { describe, it } from "node:test";
import {
  DotCheckpointEngine,
  HISTORY_NODE_KIND,
  LINE_KIND,
  STREAM_STATUS,
  TRANSPORT_KIND,
  XaiopStream,
  XaiopSyntaxError,
  parseSync,
} from "../dist/index.js";
import { chunksOf, waitStatus } from "./helpers/stream.js";

const SEED = ">\n>orders-\n>\nid:A1\nstatus:pending\n<\n>\nid:A2\nstatus:pending\n<\n>\nid:A3\nstatus:done\n.\n";
const SELECT_A2 = "@orders\n?id:A2\nstatus:shipped\n.\n";
const SPLICE_A1 = "@orders\n?id:A1\n&\n.\n";
const STAR_SHIPPED = "@orders\n?*status:shipped\nchecked:true\n.\n";
const LOCATE_SELECT = "=orders\n?1\nnote:ok\n.\n";
const FULL = SEED + SELECT_A2 + SPLICE_A1;

const AFTER_SEED = {
  orders: [
    { id: "A1", status: "pending" },
    { id: "A2", status: "pending" },
    { id: "A3", status: "done" },
  ],
};
const AFTER_SELECT = {
  orders: [
    { id: "A1", status: "pending" },
    { id: "A2", status: "shipped" },
    { id: "A3", status: "done" },
  ],
};
const AFTER_SPLICE = {
  orders: [
    { id: "A2", status: "shipped" },
    { id: "A3", status: "done" },
  ],
};
const AFTER_INTERCEPT = {
  orders: [
    { id: "A1", status: "shipped" },
    { id: "A2", status: "pending" },
    { id: "A3", status: "done" },
  ],
};
const AFTER_STAR = {
  orders: [
    { id: "A1", status: "pending" },
    { id: "A2", status: "shipped", checked: true },
    { id: "A3", status: "done" },
  ],
};

function rewriteSelectA2({ view, raw }) {
  return view.kind === LINE_KIND.SELECT && view.path === "id:A2" ? "?id:A1" : raw;
}

function makeEngine(opts = {}) {
  const chunks = [];
  const engine = new DotCheckpointEngine({
    streamProcessing: true,
    mergeChunkWindow: false,
    onChunk: (d) => chunks.push(d),
    ...opts,
  });
  return { engine, chunks };
}

describe("array select × history snapshot", () => {
  it("records after-trees for seed / predicate / splice", () => {
    const { engine } = makeEngine({ historySnapshot: true });
    engine.push(FULL);
    const h = engine.history;
    assert.equal(h.length, 3);
    assert.deepEqual(h.getAfter(0), AFTER_SEED);
    assert.deepEqual(h.getAfter(1), AFTER_SELECT);
    assert.deepEqual(h.getAfter(2), AFTER_SPLICE);
    assert.deepEqual(h.getBefore(1), AFTER_SEED);
    assert.deepEqual(h.getBefore(2), AFTER_SELECT);
    const cmp = h.compare(0, 2);
    assert.deepEqual(cmp.a, AFTER_SEED);
    assert.deepEqual(cmp.b, AFTER_SPLICE);
    assert.deepEqual(h.viewRange(0, 1).json, AFTER_SELECT);
    assert.deepEqual(h.exportTimeRoot().map((n) => n.kind), [
      HISTORY_NODE_KIND.DOT,
      HISTORY_NODE_KIND.DOT,
      HISTORY_NODE_KIND.DOT,
    ]);
  });

  it("viewRange without retained wire falls back to after[to]", () => {
    const { engine } = makeEngine({ historySnapshot: true, retainWireHistory: false });
    engine.push(FULL);
    const h = engine.history;
    assert.equal(h.getNode(1).wire, null);
    assert.deepEqual(h.viewRange(0, 1).json, AFTER_SELECT);
  });

  it("emitDiff false still records after-trees", () => {
    const { engine, chunks } = makeEngine({ historySnapshot: true, emitDiff: false });
    engine.push(FULL);
    assert.deepEqual(engine.history.getAfter(2), AFTER_SPLICE);
    assert.ok(chunks.every((d) => d == null));
    assert.equal(engine.history.getDiff(1), null);
  });

  it("compat true does not change STRICT select after-trees", () => {
    const { engine } = makeEngine({ historySnapshot: true, compat: true });
    engine.push(FULL);
    assert.deepEqual(engine.history.getAfter(2), AFTER_SPLICE);
  });

  it("eof tail after an open select phase", () => {
    const { engine } = makeEngine({ historySnapshot: true });
    engine.push(SEED + "@orders\n?id:A2\nstatus:shipped\n");
    engine.finish();
    const root = engine.history.exportTimeRoot();
    assert.deepEqual(
      root.map((n) => n.kind),
      [HISTORY_NODE_KIND.DOT, HISTORY_NODE_KIND.TAIL],
    );
    assert.deepEqual(root[1].after, AFTER_SELECT);
  });
});

describe("array select × jumpTo", () => {
  it("jumpTo before splice restores the selected write, then continues", () => {
    const { engine } = makeEngine({ historySnapshot: true, historyRealtime: true });
    engine.push(FULL);
    const h = engine.history;
    assert.equal(h.liveCursor, -1);
    assert.equal(h.canJumpTo(1), true);
    const jumped = engine.jumpTo(1);
    assert.equal(jumped.kept, 2);
    assert.equal(jumped.discarded, 1);
    assert.deepEqual(jumped.after, AFTER_SELECT);
    assert.equal(h.length, 2);
    assert.equal(h.liveCursor, 1);
    assert.deepEqual(engine.committedSnapshot, AFTER_SELECT);
    assert.equal(h.canJumpTo(1), false);
    assert.throws(() => engine.jumpTo(0), RangeError);
    engine.push(STAR_SHIPPED);
    assert.equal(h.length, 3);
    assert.deepEqual(h.getAfter(2), AFTER_STAR);
  });

  it("jumpTo(0) drops select+splice; a later matching ? continues from seed", () => {
    const { engine } = makeEngine({ historyRealtime: true });
    engine.push(FULL);
    engine.jumpTo(0);
    assert.deepEqual(engine.committedSnapshot, AFTER_SEED);
    engine.push(SELECT_A2);
    assert.deepEqual(engine.history.getAfter(1), AFTER_SELECT);
  });

  it("jumpTo(0) then unmatched ?* is a syntax error", () => {
    const { engine } = makeEngine({ historyRealtime: true });
    engine.push(FULL);
    engine.jumpTo(0);
    assert.throws(() => engine.push(STAR_SHIPPED), XaiopSyntaxError);
    assert.deepEqual(engine.history.getAfter(0), AFTER_SEED);
  });

  it("retainWire false still rebuilds committed snapshot", () => {
    const { engine } = makeEngine({
      historySnapshot: true,
      historyRealtime: true,
      retainWireHistory: false,
    });
    engine.push(FULL);
    const jumped = engine.jumpTo(1);
    assert.equal(jumped.wirePrefix, null);
    assert.deepEqual(engine.committedSnapshot, AFTER_SELECT);
    engine.push(STAR_SHIPPED);
    assert.deepEqual(engine.history.getAfter(2), AFTER_STAR);
  });

  it("jump after finish reopens the engine for more select", () => {
    const { engine } = makeEngine({ historyRealtime: true });
    engine.push(FULL);
    engine.finish();
    engine.jumpTo(1);
    engine.push(STAR_SHIPPED);
    assert.deepEqual(engine.history.getAfter(2), AFTER_STAR);
  });
});

describe("array select × intercept / window / cover", () => {
  it("intercept rewrite ?id:A2→?id:A1 is re-applied on jumpTo rebuild", () => {
    const { engine } = makeEngine({
      historySnapshot: true,
      historyRealtime: true,
      lineIntercept: rewriteSelectA2,
    });
    engine.push(SEED + SELECT_A2 + SPLICE_A1);
    assert.deepEqual(engine.history.getAfter(1), AFTER_INTERCEPT);
    engine.jumpTo(1);
    assert.deepEqual(engine.committedSnapshot, AFTER_INTERCEPT);
    engine.push("@orders\n?id:A3\nnote:x\n.\n");
    assert.deepEqual(engine.history.getAfter(2), {
      orders: [
        { id: "A1", status: "shipped" },
        { id: "A2", status: "pending" },
        { id: "A3", status: "done", note: "x" },
      ],
    });
  });

  it("skipping a select line writes Content at array level", () => {
    const { engine } = makeEngine({
      historySnapshot: true,
      lineIntercept: ({ view, raw }) =>
        view.kind === LINE_KIND.SELECT ? null : raw,
    });
    engine.push(SEED + SELECT_A2);
    assert.deepEqual(engine.history.getAfter(1), {
      orders: [
        { id: "A1", status: "pending" },
        { id: "A2", status: "pending" },
        { id: "A3", status: "done" },
        { status: "shipped" },
      ],
    });
  });

  it("mergeChunkWindow true emits one chunk and three history nodes", () => {
    const { engine, chunks } = makeEngine({
      historySnapshot: true,
      mergeChunkWindow: true,
    });
    engine.push(FULL);
    assert.equal(engine.history.length, 3);
    assert.equal(chunks.length, 1);
    assert.deepEqual(engine.history.getAfter(2), AFTER_SPLICE);
  });

  it("char-chunked ?id:A2 still lands on A2", () => {
    const { engine } = makeEngine({ historySnapshot: true });
    engine.push(SEED);
    for (const ch of SELECT_A2) engine.push(ch);
    assert.deepEqual(engine.history.getAfter(1), AFTER_SELECT);
  });

  it("cover + &orders after select matches parseSync", () => {
    const wire = SEED + SELECT_A2 + "&orders\n.\n";
    const { engine } = makeEngine({ cover: true, historySnapshot: true });
    engine.push(wire);
    engine.finish();
    assert.deepEqual(engine.snapshot, parseSync(wire));
    assert.deepEqual(engine.snapshot, {});
  });

  it("cover cannot snapshot a ? array-element Cursor before bare &", () => {
    const { engine } = makeEngine({ cover: true, historySnapshot: true });
    assert.throws(
      () => engine.push(SEED + SPLICE_A1),
      (err) =>
        err instanceof XaiopSyntaxError &&
        /cannot restore Cursor after \./.test(err.message),
    );
  });
});

describe("array select × errors and compact", () => {
  it("failed later ? keeps the prior history node", () => {
    const { engine } = makeEngine({ historySnapshot: true });
    engine.push(SEED);
    assert.throws(() => engine.push("@orders\n?99\n.\n"), XaiopSyntaxError);
    assert.equal(engine.history.length, 1);
    assert.deepEqual(engine.history.getAfter(0), AFTER_SEED);
  });

  it("compactCommitted refuses while select history is retained", () => {
    const { engine } = makeEngine({ historySnapshot: true });
    engine.push(FULL);
    assert.throws(() => engine.compactCommitted(), /history/);
    engine.compactCommitted({ dropHistory: true });
    assert.equal(engine.history.length, 0);
    assert.deepEqual(engine.committedSnapshot, AFTER_SPLICE);
  });

  it("= then ? in a later phase is a cumulative history node", () => {
    const { engine } = makeEngine({ historySnapshot: true });
    engine.push(SEED + LOCATE_SELECT);
    assert.deepEqual(engine.history.getAfter(1), {
      orders: [
        { id: "A1", status: "pending" },
        { id: "A2", status: "pending", note: "ok" },
        { id: "A3", status: "done" },
      ],
    });
  });
});

describe("array select × XaiopStream jumpTo", () => {
  it("stream history jumpTo truncates after the select write", async () => {
    const stream = new XaiopStream("raw://select-hist", {
      mergeChunkWindow: false,
      historySnapshot: true,
      historyRealtime: true,
    });
    stream.onChunk(() => {});
    stream.onDone(() => {});
    stream.send({
      transport: TRANSPORT_KIND.RAW,
      source: chunksOf(FULL),
    });
    await waitStatus(stream, STREAM_STATUS.COMPLETED);
    assert.equal(stream.history.length, 3);
    assert.deepEqual(stream.history.getAfter(2), AFTER_SPLICE);
    const jumped = stream.jumpTo(1);
    assert.equal(jumped.kept, 2);
    assert.deepEqual(stream.getCommittedSnapshot(), AFTER_SELECT);
    assert.equal(stream.history.length, 2);
  });
});

File: xaiop-sdk/java/src/test/java/io/xaiop/ArraySelectHistoryTest.java
package io.xaiop;

import static io.xaiop.Fixtures.list;
import static io.xaiop.Fixtures.map;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import io.xaiop.stream.DotCheckpointEngine;
import io.xaiop.stream.LineKind;
import io.xaiop.stream.ParseHistory;
import io.xaiop.stream.StreamStatus;
import io.xaiop.stream.Transport;
import io.xaiop.stream.XaiopStream;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

/** Protocol 0.7 Draft — {@code ?} select + bare {@code &} × history / jumpTo / cover / stream. */
class ArraySelectHistoryTest {

  private static final String SEED =
      ">\n>orders-\n>\nid:A1\nstatus:pending\n<\n>\nid:A2\nstatus:pending\n<\n>\nid:A3\nstatus:done\n.\n";
  private static final String SELECT_A2 = "@orders\n?id:A2\nstatus:shipped\n.\n";
  private static final String SPLICE_A1 = "@orders\n?id:A1\n&\n.\n";
  private static final String STAR_SHIPPED = "@orders\n?*status:shipped\nchecked:true\n.\n";
  private static final String LOCATE_SELECT = "=orders\n?1\nnote:ok\n.\n";
  private static final String FULL = SEED + SELECT_A2 + SPLICE_A1;

  private static Map<String, Object> afterSeed() {
    return map(
        "orders",
        list(
            map("id", "A1", "status", "pending"),
            map("id", "A2", "status", "pending"),
            map("id", "A3", "status", "done")));
  }

  private static Map<String, Object> afterSelect() {
    return map(
        "orders",
        list(
            map("id", "A1", "status", "pending"),
            map("id", "A2", "status", "shipped"),
            map("id", "A3", "status", "done")));
  }

  private static Map<String, Object> afterSplice() {
    return map(
        "orders",
        list(map("id", "A2", "status", "shipped"), map("id", "A3", "status", "done")));
  }

  private static Map<String, Object> afterIntercept() {
    return map(
        "orders",
        list(
            map("id", "A1", "status", "shipped"),
            map("id", "A2", "status", "pending"),
            map("id", "A3", "status", "done")));
  }

  private static Map<String, Object> afterStar() {
    return map(
        "orders",
        list(
            map("id", "A1", "status", "pending"),
            map("id", "A2", "status", "shipped", "checked", true),
            map("id", "A3", "status", "done")));
  }

  private static final class Eng {
    final DotCheckpointEngine engine;
    final List<Object> chunks;

    Eng(DotCheckpointEngine engine, List<Object> chunks) {
      this.engine = engine;
      this.chunks = chunks;
    }
  }

  private static Eng makeEngine(DotCheckpointEngine.Options extra) {
    return makeEngine(extra, false);
  }

  private static Eng makeEngine(DotCheckpointEngine.Options extra, boolean mergeChunkWindow) {
    List<Object> chunks = new ArrayList<>();
    extra.streamProcessing(true).mergeChunkWindow(mergeChunkWindow).onChunk(chunks::add);
    return new Eng(extra.build(), chunks);
  }

  private static Eng snap() {
    return makeEngine(DotCheckpointEngine.Options.builder().historySnapshot(true));
  }

  private static void waitStatus(XaiopStream stream, StreamStatus want) throws Exception {
    long deadline = System.nanoTime() + Duration.ofSeconds(5).toNanos();
    while (stream.status() != want) {
      if (stream.status() == StreamStatus.ERROR) {
        throw new AssertionError("stream error", stream.lastError());
      }
      if (System.nanoTime() > deadline) {
        throw new AssertionError("timeout waiting for " + want + ", got " + stream.status());
      }
      Thread.sleep(4);
    }
  }

  @Test
  void snapshotAfterTrees() {
    Eng e = snap();
    e.engine.push(FULL);
    ParseHistory h = e.engine.history();
    assertEquals(3, h.length());
    assertEquals(afterSeed(), h.getAfter(0));
    assertEquals(afterSelect(), h.getAfter(1));
    assertEquals(afterSplice(), h.getAfter(2));
    assertEquals(afterSeed(), h.getBefore(1));
    ParseHistory.CompareResult cmp = h.compare(0, 2);
    assertEquals(afterSeed(), cmp.a);
    assertEquals(afterSplice(), cmp.b);
    assertEquals(afterSelect(), h.viewRange(0, 1).json);
  }

  @Test
  void viewRangeWithoutWire() {
    Eng e =
        makeEngine(
            DotCheckpointEngine.Options.builder()
                .historySnapshot(true)
                .retainWireHistory(false));
    e.engine.push(FULL);
    assertNull(e.engine.history().getNode(1).wire);
    assertEquals(afterSelect(), e.engine.history().viewRange(0, 1).json);
  }

  @Test
  void emitDiffFalseStillRecordsAfter() {
    Eng e =
        makeEngine(
            DotCheckpointEngine.Options.builder().historySnapshot(true).emitDiff(false));
    e.engine.push(FULL);
    assertEquals(afterSplice(), e.engine.history().getAfter(2));
    assertTrue(e.chunks.stream().allMatch(d -> d == null));
    assertNull(e.engine.history().getDiff(1));
  }

  @Test
  void compatTrueStrictSelectTrees() {
    Eng e =
        makeEngine(DotCheckpointEngine.Options.builder().historySnapshot(true).compat(true));
    e.engine.push(FULL);
    assertEquals(afterSplice(), e.engine.history().getAfter(2));
  }

  @Test
  void eofTailOpenSelect() {
    Eng e = snap();
    e.engine.push(SEED + "@orders\n?id:A2\nstatus:shipped\n");
    e.engine.finish();
    List<ParseHistory.HistoryNode> root = e.engine.history().exportTimeRoot();
    assertEquals(ParseHistory.HISTORY_NODE_KIND.DOT, root.get(0).kind);
    assertEquals(ParseHistory.HISTORY_NODE_KIND.TAIL, root.get(1).kind);
    assertEquals(afterSelect(), root.get(1).after);
  }

  @Test
  void jumpToBeforeSpliceThenContinue() {
    Eng e =
        makeEngine(
            DotCheckpointEngine.Options.builder()
                .historySnapshot(true)
                .historyRealtime(true));
    e.engine.push(FULL);
    ParseHistory h = e.engine.history();
    assertEquals(-1, h.liveCursor());
    assertTrue(h.canJumpTo(1));
    ParseHistory.JumpResult jumped = e.engine.jumpTo(1);
    assertEquals(2, jumped.kept);
    assertEquals(1, jumped.discarded);
    assertEquals(afterSelect(), jumped.after);
    assertEquals(2, h.length());
    assertEquals(1, h.liveCursor());
    assertEquals(afterSelect(), e.engine.committedSnapshot());
    assertFalse(h.canJumpTo(1));
    assertThrows(IndexOutOfBoundsException.class, () -> e.engine.jumpTo(0));
    e.engine.push(STAR_SHIPPED);
    assertEquals(3, h.length());
    assertEquals(afterStar(), h.getAfter(2));
  }

  @Test
  void jumpToSeedThenMatchingSelect() {
    Eng e = makeEngine(DotCheckpointEngine.Options.builder().historyRealtime(true));
    e.engine.push(FULL);
    e.engine.jumpTo(0);
    assertEquals(afterSeed(), e.engine.committedSnapshot());
    e.engine.push(SELECT_A2);
    assertEquals(afterSelect(), e.engine.history().getAfter(1));
  }

  @Test
  void jumpToSeedThenUnmatchedStar() {
    Eng e = makeEngine(DotCheckpointEngine.Options.builder().historyRealtime(true));
    e.engine.push(FULL);
    e.engine.jumpTo(0);
    XaiopSyntaxError ex =
        assertThrows(XaiopSyntaxError.class, () -> e.engine.push(STAR_SHIPPED));
    assertTrue(ex.getMessage().contains("matched no array elements"));
    assertEquals(afterSeed(), e.engine.history().getAfter(0));
  }

  @Test
  void retainWireFalseJumpRebuild() {
    Eng e =
        makeEngine(
            DotCheckpointEngine.Options.builder()
                .historySnapshot(true)
                .historyRealtime(true)
                .retainWireHistory(false));
    e.engine.push(FULL);
    ParseHistory.JumpResult jumped = e.engine.jumpTo(1);
    assertNull(jumped.wirePrefix);
    assertEquals(afterSelect(), e.engine.committedSnapshot());
    e.engine.push(STAR_SHIPPED);
    assertEquals(afterStar(), e.engine.history().getAfter(2));
  }

  @Test
  void jumpAfterFinishReopens() {
    Eng e = makeEngine(DotCheckpointEngine.Options.builder().historyRealtime(true));
    e.engine.push(FULL);
    e.engine.finish();
    e.engine.jumpTo(1);
    e.engine.push(STAR_SHIPPED);
    assertEquals(afterStar(), e.engine.history().getAfter(2));
  }

  @Test
  void interceptRewriteReappliedOnJump() {
    Eng e =
        makeEngine(
            DotCheckpointEngine.Options.builder()
                .historySnapshot(true)
                .historyRealtime(true)
                .lineIntercept(
                    ctx ->
                        LineKind.SELECT.equals(ctx.view().kind())
                                && "id:A2".equals(ctx.view().path())
                            ? "?id:A1"
                            : ctx.raw()));
    e.engine.push(SEED + SELECT_A2 + SPLICE_A1);
    assertEquals(afterIntercept(), e.engine.history().getAfter(1));
    e.engine.jumpTo(1);
    assertEquals(afterIntercept(), e.engine.committedSnapshot());
    e.engine.push("@orders\n?id:A3\nnote:x\n.\n");
    assertEquals(
        map(
            "orders",
            list(
                map("id", "A1", "status", "shipped"),
                map("id", "A2", "status", "pending"),
                map("id", "A3", "status", "done", "note", "x"))),
        e.engine.history().getAfter(2));
  }

  @Test
  void skipSelectWritesAtArrayLevel() {
    Eng e =
        makeEngine(
            DotCheckpointEngine.Options.builder()
                .historySnapshot(true)
                .lineIntercept(
                    ctx -> LineKind.SELECT.equals(ctx.view().kind()) ? null : ctx.raw()));
    e.engine.push(SEED + SELECT_A2);
    assertEquals(
        map(
            "orders",
            list(
                map("id", "A1", "status", "pending"),
                map("id", "A2", "status", "pending"),
                map("id", "A3", "status", "done"),
                map("status", "shipped"))),
        e.engine.history().getAfter(1));
  }

  @Test
  void mergeChunkWindowOneChunkThreeNodes() {
    Eng e = makeEngine(DotCheckpointEngine.Options.builder().historySnapshot(true), true);
    e.engine.push(FULL);
    assertEquals(3, e.engine.history().length());
    assertEquals(1, e.chunks.size());
    assertEquals(afterSplice(), e.engine.history().getAfter(2));
  }

  @Test
  void charChunkedPredicate() {
    Eng e = snap();
    e.engine.push(SEED);
    for (int i = 0; i < SELECT_A2.length(); i++) {
      e.engine.push(SELECT_A2.substring(i, i + 1));
    }
    assertEquals(afterSelect(), e.engine.history().getAfter(1));
  }

  @Test
  void coverPathDeleteAfterSelect() {
    String wire = SEED + SELECT_A2 + "&orders\n.\n";
    Eng e = makeEngine(DotCheckpointEngine.Options.builder().cover(true).historySnapshot(true));
    e.engine.push(wire);
    e.engine.finish();
    assertEquals(Parse.parse(wire), e.engine.snapshot());
    assertEquals(map(), e.engine.snapshot());
  }

  @Test
  void coverCannotRestoreSelectCursorBeforeBareAmp() {
    Eng e = makeEngine(DotCheckpointEngine.Options.builder().cover(true).historySnapshot(true));
    XaiopSyntaxError ex =
        assertThrows(XaiopSyntaxError.class, () -> e.engine.push(SEED + SPLICE_A1));
    assertTrue(ex.getMessage().contains("cannot restore Cursor after ."));
  }

  @Test
  void failedLaterSelectKeepsPriorNode() {
    Eng e = snap();
    e.engine.push(SEED);
    assertThrows(XaiopSyntaxError.class, () -> e.engine.push("@orders\n?99\n.\n"));
    assertEquals(1, e.engine.history().length());
    assertEquals(afterSeed(), e.engine.history().getAfter(0));
  }

  @Test
  void compactCommittedRefusesUntilDrop() {
    Eng e = snap();
    e.engine.push(FULL);
    IllegalStateException ex =
        assertThrows(IllegalStateException.class, e.engine::compactCommitted);
    assertTrue(ex.getMessage().contains("history") || ex.getMessage().contains("indices"));
    e.engine.compactCommitted(true);
    assertEquals(0, e.engine.history().length());
    assertEquals(afterSplice(), e.engine.committedSnapshot());
  }

  @Test
  void locateThenSelectLaterPhase() {
    Eng e = snap();
    e.engine.push(SEED + LOCATE_SELECT);
    assertEquals(
        map(
            "orders",
            list(
                map("id", "A1", "status", "pending"),
                map("id", "A2", "status", "pending", "note", "ok"),
                map("id", "A3", "status", "done"))),
        e.engine.history().getAfter(1));
  }

  @Test
  void streamJumpToSelectWrite() throws Exception {
    XaiopStream stream =
        new XaiopStream(
            "raw://select-hist",
            XaiopStream.Options.defaults()
                .mergeChunkWindow(false)
                .historySnapshot(true)
                .historyRealtime(true));
    stream.onChunk(d -> {});
    stream.onDone(j -> {});
    stream.sendRaw(Transport.chunksOf(FULL));
    waitStatus(stream, StreamStatus.COMPLETED);
    ParseHistory h = stream.history();
    assertEquals(3, h.length());
    assertEquals(afterSplice(), h.getAfter(2));
    ParseHistory.JumpResult jumped = stream.jumpTo(1);
    assertEquals(2, jumped.kept);
    assertEquals(afterSelect(), stream.getCommittedSnapshot());
    assertEquals(2, stream.history().length());
  }
}
